go-controller/pkg/metrics/master.go

```diff
@@ -477,99 +507,239 @@ const (
+type podEvent struct {
+	timestamp time.Time
+	old       model.Model
+	new       model.Model
+	uid       kapimtypes.UID
+}
+
+type configEvent struct {
+	kindStartTimestamp int
+	txStartTimestamp   int
+	kind               string
+	cfg                int
+}
+
 type item struct {
-	op     operation
-	t      time.Time
-	old    model.Model
-	new    model.Model
-	podUID kapimtypes.UID
+	op operation
+	p  *podEvent
+	c  *configEvent
 }
 
 type ControlPlaneRecorder struct {
+	// pod monitor
 	podRecords map[kapimtypes.UID]*record
-	queue      workqueue.Interface
+
+	//config duration monitor
+	mu            sync.Mutex
+	enabled       bool
+	available     bool
+	reportKind    map[string]prometheus.Histogram
+	configRecords []*configEvent
+
+	queue workqueue.Interface
 }
 
+var cpr *ControlPlaneRecorder
+
 func NewControlPlaneRecorder() *ControlPlaneRecorder {
-	return &ControlPlaneRecorder{
-		podRecords: make(map[kapimtypes.UID]*record),
+	cpr = &ControlPlaneRecorder{}
+	return cpr
+}
+
+func GetControlPlaneRecorder() *ControlPlaneRecorder {
+	if cpr == nil {
+		klog.Warning("Attempted to retrieve before NewControlPlaneRecorder() is called")
+		cpr = &ControlPlaneRecorder{}
 	}
+	if cpr.queue == nil {
+		klog.Warning("Retrieving control plane recorder but it is not Run()")
+	}
+	return cpr
 }
 
-//Run will manage metrics for monitoring control plane objects and block until stop signal received.
-//Monitors pod setup latency
-func (cpr *ControlPlaneRecorder) Run(sbClient libovsdbclient.Client, stop <-chan struct{}) {
-	// only register the metrics when we want them
+//Run monitors pod setup latency and optionally, the duration for OVN-Kube master to configure k8 kinds
+func (cpr *ControlPlaneRecorder) Run(sbClient, nbClient libovsdbclient.Client, kube kube.Interface, stop <-chan struct{}) {
+	cpr.queue = workqueue.New()
+
+	go func() {
+		wait.Until(cpr.runWorker, queueCheckPeriod, stop)
+		cpr.queue.ShutDown()
+	}()
+
+	// configure pod monitor
 	prometheus.MustRegister(metricFirstSeenLSPLatency)
 	prometheus.MustRegister(metricLSPPortBindingLatency)
 	prometheus.MustRegister(metricPortBindingUpLatency)
 	prometheus.MustRegister(metricPortBindingChassisLatency)
-	cpr.queue = workqueue.New()
+	cpr.podRecords = make(map[kapimtypes.UID]*record)
 
 	sbClient.Cache().AddEventHandler(&cache.EventHandlerFuncs{
 		AddFunc: func(table string, model model.Model) {
 			if table != portBindingTable {
 				return
 			}
-			cpr.queue.Add(item{op: addPortBinding, old: model, t: time.Now()})
+			if !cpr.queueFull() {
+				cpr.queue.Add(&item{addPortBinding, &podEvent{old: model, timestamp: time.Now()}, nil})
+			}
 		},
 		UpdateFunc: func(table string, old model.Model, new model.Model) {
 			if table != portBindingTable {
 				return
 			}
-			cpr.queue.Add(item{op: updatePortBinding, old: old, new: new, t: time.Now()})
-		},
-		DeleteFunc: func(table string, model model.Model) {
+			if !cpr.queueFull() {
```

Comment:
since for `PortBinding` we only care about changes for `Chassis` and `Up` columns, will it make sense to check here if one of these values has changed here before putting new item to the queue?

Reply:
done